internal/pkg/ntp/ntp.go

```diff
@@ -0,0 +1,307 @@
+func (syncer *Syncer) EpochChange() <-chan struct{} {
+	return syncer.epochChangeCh
+}
+
+func (syncer *Syncer) getTimeServers() []string {
+	syncer.timeServersMu.Lock()
+	defer syncer.timeServersMu.Unlock()
+
+	return syncer.timeServers
+}
+
+func (syncer *Syncer) getLastSyncServer() string {
+	syncer.timeServersMu.Lock()
+	defer syncer.timeServersMu.Unlock()
+
+	return syncer.lastSyncServer
+}
+
+func (syncer *Syncer) setLastSyncServer(lastSyncServer string) {
+	syncer.timeServersMu.Lock()
+	defer syncer.timeServersMu.Unlock()
+
+	syncer.lastSyncServer = lastSyncServer
+}
+
+// SetTimeServers sets the list of time servers to use.
+func (syncer *Syncer) SetTimeServers(timeServers []string) {
+	syncer.timeServersMu.Lock()
+	defer syncer.timeServersMu.Unlock()
+
+	if reflect.DeepEqual(timeServers, syncer.timeServers) {
+		return
+	}
+
+	syncer.timeServers = append([]string(nil), timeServers...)
+	syncer.lastSyncServer = ""
+
+	syncer.restartSync()
+}
+
+func (syncer *Syncer) restartSync() {
+	select {
+	case syncer.restartSyncCh <- struct{}{}:
+	default:
+	}
+}
+
+// Run runs the sync process.
+//
+// Run is usually run in a goroutine.
+// When context is canceled, sync process aborts.
+//
+//nolint: gocyclo
+func (syncer *Syncer) Run(ctx context.Context) {
+	for {
+		lastSyncServer, resp, err := syncer.query(ctx)
+		if err != nil {
+			return
+		}
+
+		if resp != nil && resp.Validate() == nil {
+			if err = syncer.adjustTime(resp.ClockOffset, lastSyncServer); err != nil {
+				syncer.logger.Printf("error adjusting time: %s", err)
+			} else if !syncer.timeSyncNotified {
+				// successful first time sync, notify about it
+				close(syncer.timeSynced)
+
+				syncer.timeSyncNotified = true
+			}
+		}
+
+		// Set some variance with how frequently we poll ntp servers.
+		// This is based on rand(MaxPoll) + MinPoll so we wait at least
+		// MinPoll.
+		sleepInterval := time.Duration(rand.Intn(int(syncer.maxPoll.Seconds())))*time.Second + syncer.minPoll
+
+		if resp == nil {
+			// if no response was ever received, consider doing short sleep to retry sooner as it's not KISS of death response
+			sleepInterval = syncer.minPoll / 2
+		}
+
+		select {
+		case <-ctx.Done():
+			return
+		case <-syncer.restartSyncCh:
+			// time servers got changed, restart the loop immediately
+		case <-time.After(sleepInterval):
+		}
+	}
+}
+
+func (syncer *Syncer) query(ctx context.Context) (lastSyncServer string, resp *ntp.Response, err error) {
+	lastSyncServer = syncer.getLastSyncServer()
+
+	if lastSyncServer != "" {
+		resp, err = syncer.queryServer(lastSyncServer)
+		if err != nil {
+			syncer.logger.Printf("ntp query error with server %q: %v", lastSyncServer, err)
+
+			lastSyncServer = ""
+			err = nil
+		}
+	}
+
+	if lastSyncServer == "" {
+		var serverList []string
+
+		serverList, err = syncer.resolveServers(ctx)
+		if err != nil {
+			return lastSyncServer, resp, err
+		}
+
+		for _, server := range serverList {
```

Comment:
Not a big deal, but I wonder if it should skip previous `lastSyncServer` which has just failed.

Reply:
yep, good point, I haven't thought about it

Reply:
fixed!